In [1]:
from pathlib import Path

import numpy as np
import pandas as pd

In [2]:
DATA_PATH = Path("../data/raw/bank.csv")

df = pd.read_csv(DATA_PATH)

df.head()

,age,job,marital,education,default,balance,housing,loan,contact,day,month,duration,campaign,pdays,previous,poutcome,deposit
0,59,admin.,married,secondary,no,2343,yes,no,unknown,5,may,1042,1,-1,0,unknown,yes
1,56,admin.,married,secondary,no,45,no,no,unknown,5,may,1467,1,-1,0,unknown,yes
2,41,technician,married,secondary,no,1270,yes,no,unknown,5,may,1389,1,-1,0,unknown,yes
3,55,services,married,secondary,no,2476,yes,no,unknown,5,may,579,1,-1,0,unknown,yes
4,54,admin.,married,tertiary,no,184,no,no,unknown,5,may,673,2,-1,0,unknown,yes


In [4]:
df.shape

(11162, 17)

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 11162 entries, 0 to 11161
Data columns (total 17 columns):
 #   Column     Non-Null Count  Dtype
---  ------     --------------  -----
 0   age        11162 non-null  int64
 1   job        11162 non-null  str  
 2   marital    11162 non-null  str  
 3   education  11162 non-null  str  
 4   default    11162 non-null  str  
 5   balance    11162 non-null  int64
 6   housing    11162 non-null  str  
 7   loan       11162 non-null  str  
 8   contact    11162 non-null  str  
 9   day        11162 non-null  int64
 10  month      11162 non-null  str  
 11  duration   11162 non-null  int64
 12  campaign   11162 non-null  int64
 13  pdays      11162 non-null  int64
 14  previous   11162 non-null  int64
 15  poutcome   11162 non-null  str  
 16  deposit    11162 non-null  str  
dtypes: int64(7), str(10)
memory usage: 1.4 MB


In [6]:
df.columns.tolist()

['age',
 'job',
 'marital',
 'education',
 'default',
 'balance',
 'housing',
 'loan',
 'contact',
 'day',
 'month',
 'duration',
 'campaign',
 'pdays',
 'previous',
 'poutcome',
 'deposit']

In [7]:
df.isna().sum().sort_values(ascending=False)

age          0
job          0
marital      0
education    0
default      0
balance      0
housing      0
loan         0
contact      0
day          0
month        0
duration     0
campaign     0
pdays        0
previous     0
poutcome     0
deposit      0
dtype: int64

In [8]:
df.duplicated().sum()

np.int64(0)

In [9]:
df.nunique().sort_values()

housing         2
loan            2
default         2
deposit         2
contact         3
marital         3
poutcome        4
education       4
month          12
job            12
day            31
previous       34
campaign       36
age            76
pdays         472
duration     1428
balance      3805
dtype: int64

In [10]:
for col in df.columns:
    if df[col].nunique() <= 20:
        print(f"\n=== {col} ===")
        print(df[col].value_counts(dropna=False))


=== job ===
job
management       2566
blue-collar      1944
technician       1823
admin.           1334
services          923
retired           778
self-employed     405
student           360
unemployed        357
entrepreneur      328
housemaid         274
unknown            70
Name: count, dtype: int64

=== marital ===
marital
married     6351
single      3518
divorced    1293
Name: count, dtype: int64

=== education ===
education
secondary    5476
tertiary     3689
primary      1500
unknown       497
Name: count, dtype: int64

=== default ===
default
no     10994
yes      168
Name: count, dtype: int64

=== housing ===
housing
no     5881
yes    5281
Name: count, dtype: int64

=== loan ===
loan
no     9702
yes    1460
Name: count, dtype: int64

=== contact ===
contact
cellular     8042
unknown      2346
telephone     774
Name: count, dtype: int64

=== month ===
month
may    2824
aug    1519
jul    1514
jun    1222
nov     943
apr     923
feb     776
oct     392
jan     344
sep     3

In [11]:
possible_targets = ["deposit", "y", "subscribed", "target", "response"]

for col in possible_targets:
    if col in df.columns:
        print(f"\nPossible target: {col}")
        print(df[col].value_counts(dropna=False))
        print(df[col].value_counts(normalize=True, dropna=False))


Possible target: deposit
deposit
no     5873
yes    5289
Name: count, dtype: int64
deposit
no     0.52616
yes    0.47384
Name: proportion, dtype: float64


Ячейка 11 — numeric / categorical columns

In [12]:
target_col = "deposit"

numeric_cols = df.drop(columns=[target_col]).select_dtypes(include=["number"]).columns.tolist()
categorical_cols = df.drop(columns=[target_col]).select_dtypes(exclude=["number"]).columns.tolist()

numeric_cols, categorical_cols

(['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous'],
 ['job',
  'marital',
  'education',
  'default',
  'housing',
  'loan',
  'contact',
  'month',
  'poutcome'])

Ячейка 12 — cardinality

In [13]:
df[categorical_cols].nunique().sort_values(ascending=False)

job          12
month        12
education     4
poutcome      4
marital       3
contact       3
default       2
housing       2
loan          2
dtype: int64

Ячейка 13 — value_counts по категориальным признакам

In [14]:
for col in categorical_cols + [target_col]:
    print(f"\n=== {col} ===")
    print(df[col].value_counts(dropna=False))


=== job ===
job
management       2566
blue-collar      1944
technician       1823
admin.           1334
services          923
retired           778
self-employed     405
student           360
unemployed        357
entrepreneur      328
housemaid         274
unknown            70
Name: count, dtype: int64

=== marital ===
marital
married     6351
single      3518
divorced    1293
Name: count, dtype: int64

=== education ===
education
secondary    5476
tertiary     3689
primary      1500
unknown       497
Name: count, dtype: int64

=== default ===
default
no     10994
yes      168
Name: count, dtype: int64

=== housing ===
housing
no     5881
yes    5281
Name: count, dtype: int64

=== loan ===
loan
no     9702
yes    1460
Name: count, dtype: int64

=== contact ===
contact
cellular     8042
unknown      2346
telephone     774
Name: count, dtype: int64

=== month ===
month
may    2824
aug    1519
jul    1514
jun    1222
nov     943
apr     923
feb     776
oct     392
jan     344
sep     3

Ячейка 14 — поиск "unknown"

In [15]:
unknown_counts = {}

for col in categorical_cols:
    unknown_counts[col] = (df[col] == "unknown").sum()

pd.Series(unknown_counts).sort_values(ascending=False)

poutcome     8326
contact      2346
education     497
job            70
marital         0
housing         0
default         0
loan            0
month           0
dtype: int64

Ячейка 15 — числовые признаки

In [16]:
df[numeric_cols].describe().T

,count,mean,std,min,25%,50%,75%,max
age,11162.0,41.231948,11.913369,18.0,32.0,39.0,49.00,95.0
balance,11162.0,1528.538524,3225.413326,-6847.0,122.0,550.0,1708.00,81204.0
day,11162.0,15.658036,8.420740,1.0,8.0,15.0,22.00,31.0
duration,11162.0,371.993818,347.128386,2.0,138.0,255.0,496.00,3881.0
campaign,11162.0,2.508421,2.722077,1.0,1.0,2.0,3.00,63.0
pdays,11162.0,51.330407,108.758282,-1.0,-1.0,-1.0,20.75,854.0
previous,11162.0,0.832557,2.292007,0.0,0.0,0.0,1.00,58.0


Ячейка 16 — date-like признаки

In [17]:
df[["day", "month"]].head(20)

,day,month
0,5,may
1,5,may
2,5,may
3,5,may
4,5,may
5,5,may
6,6,may
7,6,may
8,6,may
9,6,may


In [18]:
df["month"].value_counts().sort_index()

month
apr     923
aug    1519
dec     110
feb     776
jan     344
jul    1514
jun    1222
mar     276
may    2824
nov     943
oct     392
sep     319
Name: count, dtype: int64

In [19]:
df["day"].describe()

count    11162.000000
mean        15.658036
std          8.420740
min          1.000000
25%          8.000000
50%         15.000000
75%         22.000000
max         31.000000
Name: day, dtype: float64

Ячейка 17 — leakage-risk признаки

In [20]:
leakage_risk_cols = ["duration", "pdays", "previous", "poutcome"]

for col in leakage_risk_cols:
    if col in df.columns:
        print(f"\n=== {col} ===")
        print("nunique:", df[col].nunique())
        print(df[col].value_counts(dropna=False).head(20))


=== duration ===
nunique: 1428
duration
97     39
161    38
158    37
150    37
119    37
144    36
114    36
112    35
131    35
90     35
136    35
135    34
152    33
126    33
185    33
101    33
139    33
125    32
193    32
109    32
Name: count, dtype: int64

=== pdays ===
nunique: 472
pdays
-1      8324
 92      106
 182      89
 91       84
 181      81
 183      73
 184      52
 94       42
 93       41
 95       39
 87       34
 90       32
 98       31
 185      30
 187      28
 96       26
 97       26
 189      24
 188      24
 99       22
Name: count, dtype: int64

=== previous ===
nunique: 34
previous
0     8324
1      887
2      693
3      435
4      244
5      165
6      117
7       77
8       60
9       34
10      30
11      23
12      16
13      11
17      11
14       6
15       5
19       4
23       2
20       2
Name: count, dtype: int64

=== poutcome ===
nunique: 4
poutcome
unknown    8326
failure    1228
success    1071
other       537
Name: count, dtype: int64


Ячейка 18 — target encoding preview, только для аудита

Важно: это не preprocessing, а просто проверка распределения target.

In [21]:
df[target_col].value_counts()

deposit
no     5873
yes    5289
Name: count, dtype: int64

In [22]:
df[target_col].value_counts(normalize=True)

deposit
no     0.52616
yes    0.47384
Name: proportion, dtype: float64

Что должно получиться

После этого у нас будет почти полный Stage 0 audit:

список numeric columns;
список categorical columns;
распределения категорий;
количество "unknown";
понимание, что day и month — это временные признаки, но не полноценная дата;
предварительная классификация leakage-risk колонок;
первая бизнес-постановка задачи.

Ячейка — numeric/categorical columns

In [23]:
target_col = "deposit"

numeric_cols = df.drop(columns=[target_col]).select_dtypes(include=["number"]).columns.tolist()
categorical_cols = df.drop(columns=[target_col]).select_dtypes(exclude=["number"]).columns.tolist()

print("Numeric columns:")
print(numeric_cols)

print("\nCategorical columns:")
print(categorical_cols)

Numeric columns:
['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']

Categorical columns:
['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome']


In [24]:
df[categorical_cols].nunique().sort_values(ascending=False)

job          12
month        12
education     4
poutcome      4
marital       3
contact       3
default       2
housing       2
loan          2
dtype: int64

In [25]:
key_categorical_cols = [
    "job",
    "education",
    "contact",
    "month",
    "poutcome",
    "deposit",
]

for col in key_categorical_cols:
    print(f"\n=== {col} ===")
    print(df[col].value_counts(dropna=False))
    print("\nProportions:")
    print(df[col].value_counts(normalize=True, dropna=False))


=== job ===
job
management       2566
blue-collar      1944
technician       1823
admin.           1334
services          923
retired           778
self-employed     405
student           360
unemployed        357
entrepreneur      328
housemaid         274
unknown            70
Name: count, dtype: int64

Proportions:
job
management       0.229887
blue-collar      0.174162
technician       0.163322
admin.           0.119513
services         0.082691
retired          0.069701
self-employed    0.036284
student          0.032252
unemployed       0.031984
entrepreneur     0.029385
housemaid        0.024548
unknown          0.006271
Name: proportion, dtype: float64

=== education ===
education
secondary    5476
tertiary     3689
primary      1500
unknown       497
Name: count, dtype: int64

Proportions:
education
secondary    0.490593
tertiary     0.330496
primary      0.134385
unknown      0.044526
Name: proportion, dtype: float64

=== contact ===
contact
cellular     8042
unknown      23

Ячейка — month/day target relationship, только аудит

Это не обучение модели, а проверка временной структуры и распределений.

In [26]:
pd.crosstab(df["month"], df[target_col], normalize="index").sort_index()

deposit,no,yes
month,,
apr,0.374865,0.625135
aug,0.547070,0.452930
dec,0.090909,0.909091
feb,0.431701,0.568299
jan,0.587209,0.412791
jul,0.585865,0.414135
jun,0.553191,0.446809
mar,0.101449,0.898551
may,0.672450,0.327550


In [27]:
pd.crosstab(df["day"], df[target_col], normalize="index").head()

deposit,no,yes
day,,
1,0.262295,0.737705
2,0.455090,0.544910
3,0.418301,0.581699
4,0.427861,0.572139
5,0.549266,0.450734


In [28]:
df["month"].value_counts().sort_index()

month
apr     923
aug    1519
dec     110
feb     776
jan     344
jul    1514
jun    1222
mar     276
may    2824
nov     943
oct     392
sep     319
Name: count, dtype: int64

In [29]:
df["day"].value_counts().sort_index()

day
1     122
2     334
3     306
4     402
5     477
6     447
7     382
8     419
9     364
10    163
11    373
12    445
13    453
14    463
15    466
16    369
17    411
18    548
19    384
20    570
21    452
22    269
23    245
24    122
25    224
26    252
27    284
28    410
29    388
30    478
31    140
Name: count, dtype: int64

Ячейка — proposed feature groups

In [30]:
target_col = "deposit"

leakage_risk_cols = ["duration"]

date_like_cols = ["day", "month"]

history_campaign_cols = ["pdays", "previous", "poutcome"]

candidate_feature_cols = [
    col for col in df.columns 
    if col != target_col and col not in leakage_risk_cols
]

candidate_feature_cols

['age',
 'job',
 'marital',
 'education',
 'default',
 'balance',
 'housing',
 'loan',
 'contact',
 'day',
 'month',
 'campaign',
 'pdays',
 'previous',
 'poutcome']

Ячейка — feature type lists

In [31]:
target_col = "deposit"

numeric_cols = df.drop(columns=[target_col]).select_dtypes(include=["number"]).columns.tolist()
categorical_cols = df.drop(columns=[target_col]).select_dtypes(exclude=["number"]).columns.tolist()

print("Numeric columns:")
print(numeric_cols)

print("\nCategorical columns:")
print(categorical_cols)

Numeric columns:
['age', 'balance', 'day', 'duration', 'campaign', 'pdays', 'previous']

Categorical columns:
['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome']


In [ ]:
Ячейка — target rate по duration

In [32]:
df.groupby(target_col)["duration"].describe()

,count,mean,std,min,25%,50%,75%,max
deposit,,,,,,,,
no,5873.0,223.130257,208.577530,2.0,94.0,163.0,282.0,3284.0
yes,5289.0,537.294574,392.525262,8.0,244.0,426.0,725.0,3881.0


In [ ]:
Ячейка — target rate по poutcome

In [33]:
pd.crosstab(df["poutcome"], df[target_col], normalize="index")

deposit,no,yes
poutcome,,
failure,0.496743,0.503257
other,0.428305,0.571695
success,0.086835,0.913165
unknown,0.593322,0.406678


Ячейка — target rate по contact

In [34]:
pd.crosstab(df["contact"], df[target_col], normalize="index")

deposit,no,yes
contact,,
cellular,0.456727,0.543273
telephone,0.496124,0.503876
unknown,0.774084,0.225916


Ячейка — target rate по month

In [35]:
pd.crosstab(df["month"], df[target_col], normalize="index").sort_index()

deposit,no,yes
month,,
apr,0.374865,0.625135
aug,0.547070,0.452930
dec,0.090909,0.909091
feb,0.431701,0.568299
jan,0.587209,0.412791
jul,0.585865,0.414135
jun,0.553191,0.446809
mar,0.101449,0.898551
may,0.672450,0.327550


Ячейка — preliminary feature decision

In [36]:
leakage_excluded_cols = ["duration"]

candidate_feature_cols = [
    col for col in df.columns
    if col != target_col and col not in leakage_excluded_cols
]

candidate_numeric_cols = [
    col for col in numeric_cols
    if col in candidate_feature_cols
]

candidate_categorical_cols = [
    col for col in categorical_cols
    if col in candidate_feature_cols
]

print("Target:", target_col)
print("Excluded due to leakage risk:", leakage_excluded_cols)

print("\nCandidate numeric columns:")
print(candidate_numeric_cols)

print("\nCandidate categorical columns:")
print(candidate_categorical_cols)

Target: deposit
Excluded due to leakage risk: ['duration']

Candidate numeric columns:
['age', 'balance', 'day', 'campaign', 'pdays', 'previous']

Candidate categorical columns:
['job', 'marital', 'education', 'default', 'housing', 'loan', 'contact', 'month', 'poutcome']


## Stage 0 preliminary task framing

Project: 04_bank_marketing

Problem type: binary classification.

Target column: `deposit`.

Positive class: `yes`.

Initial business framing:
Predict whether a client will subscribe to a term deposit before the final outcome of the marketing contact is known.

Dataset size:
11162 rows, 17 columns.

Class distribution:
- no: 5873 / 52.6%
- yes: 5289 / 47.4%

Main metrics:
- PR-AUC
- F1
- precision
- recall

Secondary metrics:
- ROC-AUC
- accuracy as auxiliary only

Leakage risks:
- `duration` is high-risk for pre-contact prediction because call duration is known only after/during the contact.
- `pdays`, `previous`, and `poutcome` describe previous campaign history. They are not automatically leakage if this information is known before the current campaign contact.

Temporal validation:
- `day` and `month` exist, but there is no year or full timestamp.
- A fully reliable chronological split is not possible from available columns alone.
- Initial plan: stratified train/test split, with test set used only once at final evaluation.

## Stage 0 — Dataset audit and task framing

Project: `04_bank_marketing`

Dataset: Bank Marketing Dataset

### Dataset size

- Rows: 11162
- Columns: 17

### Target

Target column: `deposit`

Positive class: `yes`

Class distribution:

- `no`: 5873 / 52.6%
- `yes`: 5289 / 47.4%

There is no severe class imbalance, but this is still a business classification problem where precision/recall trade-off matters.

### Problem type

Binary classification.

### Initial business framing

Predict whether a client will subscribe to a term deposit before the final outcome of the marketing contact is known.

This framing is useful for campaign prioritization: the model can help decide which clients are more promising to contact.

### Numeric columns

Original numeric columns:

- `age`
- `balance`
- `day`
- `duration`
- `campaign`
- `pdays`
- `previous`

Candidate numeric columns:

- `age`
- `balance`
- `day`
- `campaign`
- `pdays`
- `previous`

### Categorical columns

Candidate categorical columns:

- `job`
- `marital`
- `education`
- `default`
- `housing`
- `loan`
- `contact`
- `month`
- `poutcome`

### Missing values

No explicit NaN values were found.

However, several categorical columns contain `unknown` values:

- `poutcome`: 8326
- `contact`: 2346
- `education`: 497
- `job`: 70

For the initial pipeline, `unknown` will be treated as a valid category and encoded with OneHotEncoder.

### Duplicates

Duplicate rows: 0

### Date-like columns

The dataset contains:

- `day`
- `month`

There is no full date, year, timestamp, or campaign id. Because of that, a fully reliable chronological split is not possible from the available columns alone.

Initial validation plan:

- use stratified train/test split;
- reserve test set for final evaluation only;
- use cross-validation on train during model comparison and tuning.

### Leakage risks

`duration` is a high-risk leakage feature for the pre-contact business framing.

Reason: call duration is only known during or after the call. If the model is supposed to prioritize clients before contact, `duration` is unavailable at prediction time.

Decision:

- exclude `duration` from candidate features.

`pdays`, `previous`, and `poutcome` describe previous campaign history. They are not automatically leakage if this information is available before the current campaign contact.

Decision:

- keep `pdays`, `previous`, and `poutcome` as candidate features for now.

### Initial metrics

Primary metrics:

- PR-AUC
- F1
- precision
- recall

Secondary metrics:

- ROC-AUC
- accuracy as auxiliary only

Accuracy should not be the main metric because business decisions depend on the cost of false positives and false negatives.

### Main business trade-off

False positive:

- bank contacts a client who will not subscribe;
- cost: wasted call/sales effort, possible customer irritation.

False negative:

- bank does not prioritize a client who would have subscribed;
- cost: missed business opportunity.

Threshold tuning must be done only on train/CV, never on the final test set.